File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/playback/service/SeekableSamplesSource.java
package app.zxtune.playback.service;

import app.zxtune.TimeStamp;
import app.zxtune.core.Player;
import app.zxtune.core.Properties;
import app.zxtune.sound.SamplesSource;

import java.util.concurrent.TimeUnit;

class SeekableSamplesSource implements SamplesSource {

  private final Player player;

  SeekableSamplesSource(Player player) {
    this.player = player;
    player.setPosition(TimeStamp.EMPTY);
  }

  @Override
  public boolean getSamples(short[] buf) {
    if (player.render(buf)) {
      return true;
    } else {
      player.setPosition(TimeStamp.EMPTY);
      return false;
    }
  }

  @Override
  public TimeStamp getPosition() {
    return player.getPosition();
  }

  @Override
  public void setPosition(TimeStamp pos) {
    player.setPosition(pos);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/playback/service/SeekableSamplesSource.kt
package app.zxtune.playback.service

import app.zxtune.TimeStamp
import app.zxtune.core.Player
import app.zxtune.sound.SamplesSource

internal class SeekableSamplesSource(private val player: Player) : SamplesSource {

    init {
        player.position = TimeStamp.EMPTY
    }

    override fun getSamples(buf: ShortArray) = if (player.render(buf)) {
        true
    } else {
        player.position = TimeStamp.EMPTY
        false
    }

    override var position: TimeStamp by player::position
}
